src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -92,4 +102,12 @@ public void close() {
       myWorkerFuture.cancel(true);
```

Comment:
one thing I don't quite understand here: I tried to replace this method body with

```java
workerFuture = null;
workerExecutorService.shutdownNow();
```

Per the [Javadoc of `shutdownNow`](https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/util/concurrent/ExecutorService.html#shutdownNow()), this should have the same effect as the previous code, but for some reason using `shutdownNow()` here consistently causes the later call to `workerExecutorService.close()` on line 108 to block forever. This seems rather surprising, but I have no idea whether this is a bug in JRE or I'm holding it wrong, so I just reverted to the previous implementation.

If anyone has an idea what's happening here, please share.

Reply:
While the Skyframe thread is blocking on line 108, what is the worker virtual thread doing? (jstack should be able to tell)

Reply:
You want to use `jcmd ${pid} Thread.dump_to_file ${out_file}` to capture virtual thread stacks.

Reply:
If you allow me to nitpick a bit, how could one interrupt a carrier thread? My understanding is that carrier threads are OS-level functionality that's not surfaced in Java (you don't get a `Thread` object to represent the carrier thread, do you?) and interrupts are Java-level functionality which doesn't have an equivalent in the OS.

Reply:
No? Carrier threads are Java threads. Virtual threads are entirely implemented in the JRE, AFAIK, not OS-level functionality. There's even a `VirtualThread#carrierThread` field (https://github.com/openjdk/jdk21/blob/890adb6410dab4606a4f26a942aed02fb2f55387/src/java.base/share/classes/java/lang/VirtualThread.java#L131).